### PR TITLE
util/checksum: migrate to testify

### LINKED ISSUE #26169: migrate test-infra to testify for `util/checksum` pkg

### PR DESCRIPTION
**Pull request by Ranxy**
### What problem does this PR solve?

Issue Number: close #26169 

### What is changed and how it works?

What's Changed:
remove pingcap/check for utils/hack
refactor duplicate code

### Release note 

* No release note.

### CODE REVIEW

**Review by tisonkun — CHANGES_REQUESTED** [review]
Thanks for submitting this PR @Ranxy !

May you shuffle helper functions after test functions? Always we focus on test functions first and optionally take a look at its helpers.

**Comment by Ranxy** [comment]
@tisonkun Updated.

**Review by xhebox — COMMENTED** [review]
Rest LGTM

File: util/checksum/checksum_test.go [hunk 1]
```diff
@@ -596,130 +393,223 @@ func (s *testChecksumSuite) testTiCase3651and3652(c *check.C, encrypt bool) {
...
+	// This write will trigger flush.
+	n, err = w.Write([]byte("0"))
+	assert.NoError(t, err)
+	assert.Equal(t, 1, n)
+	checkFlushedData(t, f, 0, 1020, 1020, nil, buf.Bytes())
+	cacheOff := w.GetCacheDataOffset()
+	assert.Equal(t, int64(len(buf.Bytes())), cacheOff)
+}
+
+func createTestFileWithoutClose(t *testing.T, path string) (f *os.File, clean func()) {
```
[the remark this answers is not in the document]
Comment (xhebox) [reply util/checksum/checksum_test.go:485]: Maybe consider using `bytes.Buffer` with a dummy `Close`, it seems possible after my glance.

[the remark this answers is not in the document]
Comment (xhebox) [reply util/checksum/checksum_test.go:485]: Maybe consider using `bytes.Buffer` with a dummy `Close`, it seems possible after my glance.

[the remark this answers is not in the document]
Comment (tisonkun) [reply util/checksum/checksum_test.go:485]: Either is a new coupling. I'd prefer just decouple as "create", "open", "close" (return func), "remove" (return func) and compose in place. The problem here is that `defer` stmt accepts only one stmt or func block.

I'll try to create a patch on this PR in hours/

[the remark this answers is not in the document]
Comment (tisonkun) [reply util/checksum/checksum_test.go:485]: ... or make use of create tmp file.

[the remark this answers is not in the document]
Comment (Ranxy) [reply util/checksum/checksum_test.go:485]: @xhebox  This is an exciting idea, and I believe it is possible to use a fake close file for this purpose.

[the remark this answers is not in the document]
Comment (Ranxy) [reply util/checksum/checksum_test.go]: @tisonkun This sounds great, but it still requires the user to be careful about using the two return func.

[the remark this answers is not in the document]
Comment (xhebox) [reply util/checksum/checksum_test.go]: @Ranxy It is still a problem you have called multiple `createTestFile` with same paths. Fake `close` is for `bytes.Buffer` idea, which doesn't need `close` at all.

[the remark this answers is not in the document]
Comment (xhebox) [reply util/checksum/checksum_test.go]: > Either is a new coupling. I'd prefer just decouple...

I think it is acceptable to have a specific function for specific problems. The coupling is not so complex anyway.

**Comment by tisonkun** [comment]
@Ranxy I'm creating a PR for you, please wait to make other changes.

**Comment by tisonkun** [comment]
@Ranxy you can take a look at https://github.com/Ranxy/tidb/pull/1 .

**Review by tisonkun — COMMENTED** [review]
I've created a PR for you. You can take a look at https://github.com/Ranxy/tidb/pull/1 .

Basically it reduces unnecessary file operation and improve ticase comments.

File: util/checksum/checksum_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tisonkun) [reply util/checksum/checksum_test.go]: @xhebox see my patch for no file op helper at all https://github.com/Ranxy/tidb/pull/1 .

**Comment by Ranxy** [comment]
@tisonkun Your modifications are very effective, and I think @xhebox's proposal is also very good, and we can completely replace `File` with the following `bytes.buffer`. which do you think is better?

```go 
type fakeFile struct {
	buf *bytes.Buffer
}
func (f *fakeFile) Write(p []byte) (n int, err error) {
	return f.buf.Write(p)
}
func (f *fakeFile) Close() error {
	return nil
}
func (f *fakeFile) ReadAt(p []byte, off int64) (n int, err error) {
	w := f.buf.Bytes()
	lp := int64(len(p))
	lw := int64(len(w))
	if off > lw {
		return 0, io.EOF
	}
	if off+lp > lw {
		copy(p[:lw-off], w[off:])
		return int(lw - off), io.EOF
	}
	copy(p, w[off:off+lp])
	return len(p), nil
}
```

**Comment by tisonkun** [comment]
@Ranxy you can integrate two comments and I think `Close` for byte buffer is redundant.

The parallelism refactor and  ticase comment update in the PR Ranxy#1  it still valid and should be included in the final change set IMO :P

**Review by tisonkun — APPROVED** [review]
LGTM

File: util/checksum/checksum_test.go [hunk 2]
```diff
@@ -596,130 +362,213 @@ func (s *testChecksumSuite) testTiCase3651and3652(c *check.C, encrypt bool) {
...
+	lw := int64(len(w))
+	if off > lw {
+		return 0, io.EOF
+	}
+	if off+lp > lw {
+		copy(p[:lw-off], w[off:])
+		return int(lw - off), io.EOF
+	}
+	copy(p, w[off:off+lp])
+	return len(p), nil
```
Comment (xhebox) [inline util/checksum/checksum_test.go:569]: Futher simplification can be made:

```
func (f *fakeFile) ReadAt(p []byte, off int64) (n int, err error) {
	w := f.buf.Bytes()
	lp := int64(len(p))
	lw := int64(len(w))
	if off > lw {
		return 0, io.EOF
	}
	lc := copy(p, w[off:])
        if lc == lw-off {
	   return lc, io.EOF
        } else {
           return lc, nil
        }
}
```

**Review by xhebox — COMMENTED** [review]
Rest LGTM

**Review by xhebox — APPROVED** [review]

**Review by tiancaiamao — APPROVED** [review]

**Comment by Ranxy** [comment]
@tiancaiamao  @xhebox Thanks for your review! May you help with merging this PR?

File: util/checksum/checksum_test.go [hunk 3]
```diff
@@ -287,72 +162,39 @@ func (s *testChecksumSuite) testTiCase3646(c *check.C, encrypt bool) {
...
+		err := underlyingReadAt(f, encrypt, ctrCipher, 10, i*1000)
 		if err == io.EOF {
 			break
 		}
 		if i != 5 {
-			c.Assert(err, check.Equals, nil)
+			assert.NoError(t, err)
 		} else {
-			c.Assert(err, check.Equals, errChecksumFail)
+			assert.ErrorIs(t, err, errChecksumFail)
```
Comment (tisonkun) [inline util/checksum/checksum_test.go:178]: This line now creates unstable test

```
[2021-09-23T16:00:22.934Z] --- FAIL: TestModifyOneByte (0.00s)
[2021-09-23T16:00:22.934Z]     --- FAIL: TestModifyOneByte/encrypted (0.00s)
[2021-09-23T16:00:22.934Z]         checksum_test.go:179: 
[2021-09-23T16:00:22.934Z]             	Error Trace:	checksum_test.go:179
[2021-09-23T16:00:22.934Z]             	            				checksum_test.go:150
[2021-09-23T16:00:22.934Z]             	Error:      	Target error should be in err chain:
[2021-09-23T16:00:22.934Z]             	            	expected: "error checksum"
[2021-09-23T16:00:22.934Z]             	            	in chain: 
[2021-09-23T16:00:22.934Z]             	Test:       	TestModifyOneByte/encrypted
```

**Pull request merged**